core/src/apps/cardano/helpers/tx_builder.py

```diff
@@ -0,0 +1,194 @@
+        self.serializer = cbor.encode_streamed(self.state[0])
+        self.hashfn = hashlib.blake2b(outlen=32)
+        self._update_tx_hash()
+        self.keychain = keychain
+        self.msg = msg
+        self.expected_item_kind: Optional[int] = None
+
+    def _update_tx_hash(self) -> None:
+        while (chunk := next(self.serializer)) is not None:
+            self.hashfn.update(chunk)
+
+    def get_tx_hash(self) -> bytes:
+        assert self.state[0].is_filled(), "The hash calculation is not finished yet"
+        return self.hashfn.digest()
+
+    def _start_collection_at_key(
+        self,
+        key: Any,
+        collection: LazyCollection,
+        next_kind: int,
+        current_kind: Optional[int] = None,
+    ) -> None:
+        self._add_item((key, collection), current_kind)
+        self.state.append(collection)
+        self.expected_item_kind = next_kind
+
+    def _add_item(self, item: Any, item_kind: Optional[int] = None) -> None:
+        assert self.expected_item_kind == item_kind, "Item of unexpected kind added"
+        self.state[-1].append_item(item)
+        self._update_tx_hash()
+
+    def _finish_collection(
+        self, current_kind: int, next_kind: Optional[int] = None
+    ) -> None:
+        assert (
+            self.expected_item_kind == current_kind
+        ), "Tried to finish a collection of unexpected kind"
+        self.expected_item_kind = next_kind
+        assert self.state[-1].is_filled(), "The collection is not filled yet"
+        self.state.pop()
+        assert next(self.serializer) is None
+
+    def start_inputs(self) -> None:
+        self._start_collection_at_key(
+            0, LazyCollection(LIST, self.msg.inputs_count), INPUT
+        )
+
+    def add_input(self, input: CardanoTxInput) -> None:
+        self._add_item((input.prev_hash, input.prev_index), INPUT)
+
+    def finish_inputs(self) -> None:
+        self._finish_collection(INPUT)
+
+    def start_outputs(self) -> None:
+        self._start_collection_at_key(
+            1, LazyCollection(LIST, self.msg.outputs_count), OUTPUT
+        )
+
+    def add_output(self, output: CardanoTxOutput) -> None:
+        address = _get_output_address(
+            self.keychain, self.msg.protocol_magic, self.msg.network_id, output
+        )
+        if output.asset_groups_count == 0:
+            self._add_item((address, output.amount), OUTPUT)
+        else:
+            asset_groups = LazyCollection(DICT, output.asset_groups_count)
+            self._add_item((address, (output.amount, asset_groups)), OUTPUT)
+            self.state.append(asset_groups)
+            self.expected_item_kind = ASSET_GROUP
+
+    def finish_outputs(self) -> None:
+        self._finish_collection(OUTPUT)
+
+    def add_asset_group(self, asset_group: CardanoAssetGroup) -> None:
+        self._start_collection_at_key(
+            asset_group.policy_id,
+            LazyCollection(DICT, asset_group.tokens_count),
+            next_kind=TOKEN,
```

Comment:
What would we lack if we didn't have the this `next_kind` at all? It seems to me as some kind of "safe guard" but imho it's a too hard to follow, so it's kind of hard to reason about whether it works correctly or not.

If not critical for the functionality of the solution, I'd consider dropping this next_kind concept completely which I believe should simplify the logic significantly and think how to build a proper state machine aside, where the "kind transition logic" would be centralized and hence easier to follow and would offer added value in terms of guarantee of consistency as it would be easier to tell whether some illegal transition can actually happen or not.

Arguably, the tx builder already keeps track at least of the counts of the items to add, so it's IMHO reasonable to not focus on introducing such state machine right now.

Of course, I may have just misunderstood this next_kind concept and it's either actually essential or easy to follow, and I'm just biased by Ledger's state machine paradigm

Reply:
So the safety checks currently work like this:
- we assert that the number of items added is not too large https://github.com/vacuumlabs/trezor-firmware/blob/701be06f9ac51b4316589ce7805a51bbcbed55e4/core/src/apps/cardano/helpers/lazy_collection.py#L35
- and that it's not too small https://github.com/vacuumlabs/trezor-firmware/blob/701be06f9ac51b4316589ce7805a51bbcbed55e4/core/src/apps/cardano/helpers/tx_builder.py#L84
- if we want to make sure that the correct kind of items are added (i.e. the inputs array isn't filled with outputs) we have to keep track of the next expected item's kind
    - at every action we assert that the expected kind matches the item kind of the action (`current_kind`)
    - and we also have to update the next expected kind (`next_kind`)
    
This kind checking _is_ actually a primitive state machine baked into the tx builder. I agree that it hurts the overall readability. We can either remove it completely and verify only the number of items, or we can extract it to a separate `TxBuilderStateMachine` class with proper actions and state transition checking defined.

Reply:
A separate `TxBuilderStateMachine` class makes sense to me